misc iptables proxy fixes

danwinship opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fixes four iptables proxier bugs:
- The `sync_proxy_rules_iptables_total` metric was off-by-one
- It was generating affinity rules for all endpoints rather than only Ready endpoints, meaning clients with affinity might have new connections get routed to a no-longer-serving endpoint.
- It was generating endpoint chains for all endpoints (whether or not they actually got used), resulting in rule bloat.
- It only output `-j REJECT` rules for services with no endpoints at all, rather than outputting it for services with no _usable_ endpoints.

There are some other known bugs which this PR does not fix.

#### Which issue(s) this PR fixes:
Fixes #106182

#### Does this PR introduce a user-facing change?
```release-note
The kube-proxy sync_proxy_rules_iptables_total metric now gives
the correct number of rules, rather than being off by one.

Fixed multiple iptables proxy regressions introduced in 1.22:

  - When using Services with SessionAffinity, client affinity for an
    endpoint now gets broken when that endpoint becomes non-ready
    (rather than continuing until the endpoint is fully deleted).

  - Traffic to a service IP now starts getting rejected (as opposed to
    merely dropped) as soon as there are no longer any *usable*
    endpoints, rather than waiting until all of the terminating
    endpoints have terminated even when those terminating endpoints
    were not being used.

  - Chains for endpoints that won't be used are no longer output to
    iptables, saving a bit of memory/time/cpu.
```

/sig network
/priority important-soon
/cc @andrewsykim @robscott

danwinship committed:
proxy/iptables: Misc improvements to unit test

The original tests here were very shy about looking at the iptables
output, and just relied on checks like "make sure there's a jump to
table X that also includes string Y somewhere in it" and stuff like
that. Whereas the newer tests were just like, "eh, here's a wall of
text, make sure the iptables output is exactly that". Although the
latter looks messier in the code, it's more precise, and it's easier
to update correctly when you change the rules. So just make all of the
tests do a check on the full iptables output.

(Note that I didn't double-check any of the output; I'm just assuming
that the output of the current iptables proxy code is actually
correct...)

Also, don't hardcode the expected number of rules in the metrics
tests, so that there's one less thing to adjust when rules change.

Also, use t.Run() in one place to get more precise errors on failure.

danwinship committed:
proxy/iptables: Add more stuff to the unit test

The external traffic policy terminating endpoints test was testing
LoadBalancer functionality against a NodePort service with no
nodePorts (or loadBalancer IPs). It managed to test what it wanted to
test, but it's kind of dubious (and we probably _shouldn't_ have been
generating the rules it was looking for since there was no way to
actually reach the XLB chains). So fix that.

Also make the terminating endpoints test use session affinity, to add
more testing for that. Also, remove the multiple copies of the same
identical Service that is used for all of the test cases in that test.

Also add a "Cluster traffic policy and no source ranges" test to
TestOverallIPTablesRulesWithMultipleServices since we weren't really
testing either of those.

Also add a test of --masquerade-all.

danwinship committed:
proxy/iptables: Fix TestOnlyLocalNodePortsNoClusterCIDR

The test got broken to not actually use "no cluster CIDR" when
LocalDetector was implemented (and the old version of the unit test
didn't check enough to actually notice this).

danwinship committed:
proxy/iptables: Improve the sorting logic in TestOverallIPTablesRulesWithMultipleServices

The test was sorting the iptables output so as to not depend on the
order that services get processed in, but this meant it wasn't
checking the relative ordering of rules (and in fact, the ordering of
the rules in the "expected" string was wrong, in a way that would
break things if the rules had actually been generated in that order).

Add a more complicated sorting function that sorts services
alphabetically while preserving the ordering of rules within each
service.

aojea commented:
is not only session affinity, is everything, right now kube-proxy doesn't consider if the endpoint is ready, as a consequence there are more issues like this

https://github.com/kubernetes/kubernetes/issues/105657

easy to reproduce with an init containers that holds the pod to be ready, the iptables rules are installed 

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deployment
  labels:
    app: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
        active: "yes"
    spec:
      initContainers:
        - name: pause-service
          image: busybox:stable
          command: ['sh', '-c', 'echo Pausing start. && sleep 100']
      containers:
      - name: test
        image: httpd
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  type: ClusterIP
  selector:
    app: test
    active: "yes"
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

aojea commented:
we also have the tests wrong

https://github.com/kubernetes/kubernetes/blob/08bf54678e2bef8b4c72c8c6d104b0bdf19a6f75/pkg/proxy/iptables/proxier_test.go#L2981-L2982

https://github.com/kubernetes/kubernetes/blob/08bf54678e2bef8b4c72c8c6d104b0bdf19a6f75/pkg/proxy/iptables/proxier_test.go#L3038-L3041

aojea commented:
hmm, it seem I was wrong about that.
The rules are there but the service is not jumping to them

aojea commented:
I think we need something like
```diff
diff --git a/pkg/proxy/iptables/proxier.go b/pkg/proxy/iptables/proxier.go
index 79bb5b6c793..0b13ce46dd0 100644
--- a/pkg/proxy/iptables/proxier.go
+++ b/pkg/proxy/iptables/proxier.go
@@ -1002,7 +1002,19 @@ func (proxier *Proxier) syncProxyRules() {
                // Service does not have conflicting configuration such as
                // externalTrafficPolicy=Local.
                allEndpoints = proxy.FilterEndpoints(allEndpoints, svcInfo, proxier.nodeLabels)
-               hasEndpoints := len(allEndpoints) > 0
+               hasEndpoints := false
+               for _, e := range allEndpoints {
+                       if e.IsReady() {
+                               hasEndpoints = true
+                               break
+                       }
+                       if utilfeature.DefaultFeatureGate.Enabled(features.ProxyTerminatingEndpoints) {
+                               if svc.NodeLocalExternal() && e.IsServing() && e.IsTerminating() {
+                                       hasEndpoints = true
+                                       break
+                               }
+                       }
+               }
 
                svcChain := svcInfo.servicePortChainName
                if hasEndpoints {
diff --git a/pkg/proxy/iptables/proxier_test.go b/pkg/proxy/iptables/proxier_test.go
index 85f40337ddf..81b142f1f25 100644
--- a/pkg/proxy/iptables/proxier_test.go
+++ b/pkg/proxy/iptables/proxier_test.go
@@ -2938,6 +2938,77 @@ COMMIT
        assert.NotEqual(t, expectedIPTablesWithSlice, fp.iptablesData.String())
 }
 
+func TestEndpointSliceE2EReject(t *testing.T) {
+       expectedIPTablesWithSlice := `*filter
+:KUBE-SERVICES - [0:0]
+:KUBE-EXTERNAL-SERVICES - [0:0]
+:KUBE-FORWARD - [0:0]
+:KUBE-NODEPORTS - [0:0]
+-A KUBE-SERVICES -m comment --comment "ns1/svc1 has no endpoints" -m tcp -p tcp -d 172.20.1.1/32 --dport 0 -j REJECT
+-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack pod source rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack pod destination rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+COMMIT
+*nat
+:KUBE-SERVICES - [0:0]
+:KUBE-NODEPORTS - [0:0]
+:KUBE-POSTROUTING - [0:0]
+:KUBE-MARK-MASQ - [0:0]
+-A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
+-A KUBE-POSTROUTING -j MARK --xor-mark 0x4000
+-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
+-A KUBE-MARK-MASQ -j MARK --or-mark 0x4000
+-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
+COMMIT
+`
+
+       ipt := iptablestest.NewFake()
+       fp := NewFakeProxier(ipt)
+               AddressType: discovery.AddressTypeIPv4,
+               Endpoints: []discovery.Endpoint{{
+                       Addresses:  []string{"10.0.1.1"},
+                       Conditions: discovery.EndpointConditions{Ready: utilpointer.BoolPtr(false)},
+                       NodeName:   utilpointer.StringPtr(testHostname),
+               }, {
+                       Addresses:  []string{"10.0.1.4"},
+                       Conditions: discovery.EndpointConditions{Ready: utilpointer.BoolPtr(false)},
+                       NodeName:   utilpointer.StringPtr("node4"),
+               }},
+       }
+
+       fp.OnEndpointSliceAdd(endpointSlice)
+       fp.syncProxyRules()
+       assert.Equal(t, expectedIPTablesWithSlice, fp.iptablesData.String())
+}
+
```

aojea commented:
/assign @aojea @thockin @robscott

aojea commented:
and this to fix https://github.com/kubernetes/kubernetes/issues/105657

```
diff --git a/pkg/proxy/endpoints.go b/pkg/proxy/endpoints.go
index 945237dd1ce..2a01b95d746 100644
--- a/pkg/proxy/endpoints.go
+++ b/pkg/proxy/endpoints.go
@@ -121,7 +121,13 @@ func (info *BaseEndpointInfo) Port() (int, error) {
 
 // Equal is part of proxy.Endpoint interface.
 func (info *BaseEndpointInfo) Equal(other Endpoint) bool {
-       return info.String() == other.String() && info.GetIsLocal() == other.GetIsLocal()
+       return info.String() == other.String() &&
+               info.GetIsLocal() == other.GetIsLocal() &&
+               info.GetNodeName() == other.GetNodeName() &&
+               info.GetZone() == other.GetZone() &&
+               info.IsReady() == other.IsReady() &&
+               info.IsTerminating() == other.IsTerminating() &&
+               info.IsServing() == other.IsServing()
 }
 
 // GetNodeName returns the NodeName for this endpoint.
@@ -561,8 +567,20 @@ func detectStaleConnections(oldEndpointsMap, newEndpointsMap EndpointsMap, stale
                        continue
                }
 
+               epReady := 0
+               for _, ep := range epList {
+                       if ep.IsReady() {
+                               epReady++
+                       }
+               }
+               oldEpReady := 0
+               for _, ep := range oldEndpointsMap[svcPortName] {
+                       if ep.IsReady() {
+                               oldEpReady++
+                       }
+               }
                // For udp service, if its backend changes from 0 to non-0. There may exist a conntrack entry that could blackhole traffic to the service.
-               if len(epList) > 0 && len(oldEndpointsMap[svcPortName]) == 0 {
+               if epReady > 0 && oldEpReady == 0 {
                        *staleServiceNames = append(*staleServiceNames, svcPortName)
                }
        }
```

ideally we should rewrite `k8s.io/kubernetes/pkg/proxy/endpoints_test.go`  to test EndpointSlices instead of Endpoints and cover this new cases about `Ready`

Dan do you want to split the load?

jayunit100 commented:
cc @knabben

danwinship commented:
> is not only session affinity, is everything

it's not _everything_...

> hmm, it seem I was wrong about that.
> The rules are there but the service is not jumping to them

Yes, because it writes out the chains before it figures out whether it's going to need to use the terminating endpoints. I have a fix for that coming later.

>     +                       if ep.IsReady() {
>     +                               epReady++

That's still not right, because it's not taking `ProxyTerminatingEndpoints` into account.

In fact, `endpoints.go` can't figure out on its own which endpoints are actually being used, because it doesn't know whether the service is `externalTrafficPolicy: Local` or not...

> Dan do you want to split the load?

I'm in the middle of refactoring the iptables `syncProxyRules` to fix `internalTrafficPolicy` (#100313). I need to think about how to rearrange stuff to get some of the fixes in without refactoring first so it will be backportable.

danwinship commented:
/hold
may not make sense to merge this if other closely-related fixes are coming imminently...

danwinship commented:
> In fact, `endpoints.go` can't figure out on its own which endpoints are actually being used, because it doesn't know whether the service is `externalTrafficPolicy: Local` or not...

And `TopologyAwareHints` screws this up as well. To compute the list of stale endpoints, you need information from both the EndpointSlices and the Services.

Except it's even worse than that, because we can now have different stale endpoints for internal and external connections:
- UDP Service X has `externalTrafficPolicy: Local` and (implicitly) `internalTrafficPolicy: Cluster`. It has both internal and external clients. `ProxyTerminatingEndpoints` is enabled.
- The sole endpoint for the service on Node Y gets deleted and its EndpointSlice becomes `Ready: false, Serving: true, Terminating: true`. It thus stops being a valid endpoint for internal (Cluster policy) traffic from Node Y, but _does not_ stop being a valid endpoint for external (Local policy) traffic that reaches Node Y.
- At this point Node Y needs to delete conntrack entries from internal sources to the cluster IP or LB IP, but needs to _not_ delete conntrack entries from external sources to the LB IP.

But we can't implement "invalidate conntrack entries from internal sources, but not conntrack entries from external sources", because the definition of "internal" and "external" is abstracted by the LocalDetector, which might distinguish internal from external using arbitrary iptables rules that have no equivalent in `conntrack -D`. (eg, matching by source interface name).

I _think_ this is only a problem when `ProxyTerminatingEndpoints` is enabled; if it's disabled then the set of endpoints for a Local-traffic-policy service will always be a strict subset of the endpoints for a Cluster-traffic-policy service, so we can do the cleanup correctly. (We _don't_ do the cleanup correctly currently, but we _could_, as opposed to in the `ProxyTerminatingEndpoints` case, where there is no easy fix at all.)

Ideas:
1. Drop `ProxyTerminatingEndpoints`, revert the associated code, go back to the drawing board...
2. Change the rules of `ProxyTerminatingEndpoints` so that it applies to `Cluster` services as well as `Local` ones so internal and external can never have disjoint endpoints. This is probably an API break, so we'd have to make it an opt-in feature on the service? (ie, you'd have to set `publishTerminatingAddresses: true` on the service)
3. We can fix the cleanup of "black hole" conntrack entries by always preceding every UDP `-j REJECT` rule with a `-j NOTRACK` rule so we don't end up creating the conntrack entries in the first place... That doesn't seem that bad in terms of either added iptables rules (most people won't have lots and lots of services with no endpoints) or added kernel filtering load (most people won't have lots and lots of clients trying to connect to services with no endpoints so the conntrack rules won't help much anyway). We can't `NOTRACK` the rules for accepted connections though, so I don't think this will completely solve the problem for us. (But maybe it's a good idea for the rejected connections anyway?)
4. We could make the proxy set conntrack labels on UDP traffic going through it, to record whether it came from an internal or external source, so that later we can use `conntrack -D --label ...` to delete conntrack entries from either only internal or only external sources. (This _seems_ like it should work based on the iptables and conntrack man pages, but I'm not 100% sure.)

danwinship commented:
/hold cancel
There don't seem to be any other places in `syncProxyRules` itself that are confusing `allEndpoints` and `readyEndpoints`, so I think this fix (as a minimal, backport-able fix) is independent of any endpoint-tracking fixes, `internalTrafficPolicy` fixes, or `ProxyTerminatingEndpoints` fixes.

aojea commented:
There is an additional bug, if a service has ONLY non-ready endpoints it doesn't insert the reject rule
https://github.com/kubernetes/kubernetes/pull/106030#issuecomment-955784550

I think we can assume that the stale/conntrack problems were solved on the previous logic, if we are able to figure out how to solve the traffic locality problem that you mention , of course

At this point, `ProxyTerminatingEndpoints` is alpha so we have some room there ...

danwinship committed:
proxy/iptables: fix all-vs-ready endpoints a bit

Filter the allEndpoints list into readyEndpoints sooner, and set
"hasEndpoints" based (mostly) on readyEndpoints, not allEndpoints (so
that, eg, we correctly generate REJECT rules for services with no
_functioning_ endpoints, even if they have unusable terminating
endpoints).

Also, write out the endpoint chains at the top of the loop when we
iterate the endpoints for the first time, rather than copying some of
the data to another set of variables and then writing them out later.
And don't write out endpoint chains that won't be used

Also, generate affinity rules only for readyEndpoints rather than
allEndpoints, so affinity gets broken correctly when an endpoint
becomes unready.

danwinship commented:
ok, repushed with (a) a bunch more unit test improvements, (b) a combined fix for the session affinity bug and the "not generating `-j REJECT` when we're supposed to" bug

danwinship committed:
proxy/iptables: Remove a no-op check

There was code to deal with endpoints that have invalid/empty IP
addresses, but EndpointSlice validation already ensures that these
can't exist.

danwinship committed:
proxy/iptables: test that we create a consistent set of iptables rules

thockin commented:
Will look at this PR momentarily, but general comments.

1) Ugh.  
2) This is why I wanted to move KEPs to completion before opening more - too many intersecting things here.
3) This *code* is too rigid and hard to follow.  I've been thinking about how to restructure it, but have not mustered enough energy to actually start.  Something like a graph with each node representing a service or an endpoint.  Build up the graph(s) for all the paths and then render the results into iptables.  I have not looked at kpng much yet, but it sounds like we'll probably end at the same set of core ideas.  The complexity of selecting endpoints is not an iptables-specific problem.

> 1) Drop ProxyTerminatingEndpoints, revert the associated code, go back to the drawing board...

My least-favorite option

> 2) Change the rules of ProxyTerminatingEndpoints so that it applies to Cluster services as well as Local ones so internal and external can never have disjoint endpoints. This is probably an API break, so we'd have to make it an opt-in feature on the service? (ie, you'd have to set publishTerminatingAddresses: true on the service)

I want this ANYWAY.  I think it's more correct.  Andrew convinced me we could leave it out at first, but maybe not?  It's not clear to me if this would be a complete solution, but I don't think it would be an API break?

> 3) We can fix the cleanup of "black hole" conntrack entries by always preceding every UDP -j REJECT rule with a -j NOTRACK rule so we don't end up creating the conntrack entries in the first place... That doesn't seem that bad in terms of either added iptables rules (most people won't have lots and lots of services with no endpoints) or added kernel filtering load (most people won't have lots and lots of clients trying to connect to services with no endpoints so the conntrack rules won't help much anyway). We can't NOTRACK the rules for accepted connections though, so I don't think this will completely solve the problem for us. (But maybe it's a good idea for the rejected connections anyway?)

Maybe I am missing something.  if we have a REJECT rule we should not be creating conntrack records anyway?

> 4) We could make the proxy set conntrack labels on UDP traffic going through it, to record whether it came from an internal or external source, so that later we can use conntrack -D --label ... to delete conntrack entries from either only internal or only external sources. (This seems like it should work based on the iptables and conntrack man pages, but I'm not 100% sure.)

This sounds good but how?  I don't see an iptables option to set that, so we'd need some async way of setting them?

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -709,6 +711,87 @@ func errorf(msg string, rules []iptablestest.Rule, t *testing.T) {
+			lines, output = moveMatchingLines(chainName, lines, output)
+		}
+
+		// Move KUBE-NODEPORTS rules for each service, sorted by service name
```
What makes this sorted?

danwinship replied:
`findAllMatches` returns the matches in sorted order, so we start with a sorted list of services and then move the NODEPORTS rules for them one by one

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -709,6 +711,87 @@ func errorf(msg string, rules []iptablestest.Rule, t *testing.T) {
+
+// sortIPTablesRules sorts `iptables-restore` output so as to not depend on the order that
+// Services get processed in, while preserving the relative ordering of related rules.
+func sortIPTablesRules(t *testing.T, ruleData string) string {
```
This function itself deserves a test.  Which means it should take an interface instead of a testing.T, probably.

thockin reviewed (commented):
This just covers commit "Improve the sorting logic..."

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1017,15 +863,31 @@ func TestClusterIPReject(t *testing.T) {
-	if !hasJump(kubeSvcRules, iptablestest.Reject, svcIP, svcPort) {
-		errorf(fmt.Sprintf("Failed to find a %v rule for service %v with no endpoints", iptablestest.Reject, svcPortName), kubeSvcRules, t)
-	}
+	expected := `*filter
```
`dedent.Dedent()` might make these more readable?

danwinship replied:
I tried it; it's not _much_ more readable, and it makes it harder to update the expected results when they change... so I think no?

I did at least fix it so you can do 

```
        expected := `
*filter
blah blah blah
```

instead of

```
        expected := `*filter
blah blah blah
```

thockin reviewed (commented):
This only covers commit "Misc improvements to unit test"

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4370,3 +4417,81 @@ COMMIT
 	}
 }
+
+func TestMasqueradeAll(t *testing.T) {
```
this could have been a different commit :)

thockin reviewed (commented):
This covers the last commit

thockin commented:
Mostly I have nits on this PR, but now we can get down to the harder problems :(

danwinship commented:
> 1. Ugh
> 3. This code is too rigid and hard to follow.

_Some_ of the ugliness is because of things that were done to reduce the number of memory allocations per loop...

> > 2. Change the rules of ProxyTerminatingEndpoints so that it applies to Cluster services as well as Local ones so internal and external can never have disjoint endpoints. This is probably an API break, so we'd have to make it an opt-in feature on the service? (ie, you'd have to set publishTerminatingAddresses: true on the service)
> 
> I want this ANYWAY. I think it's more correct. Andrew convinced me we could leave it out at first, but maybe not? It's not clear to me if this would be a complete solution, but I don't think it would be an API break?

You want which part? "ProxyTerminatingEndpoints applies to Cluster services too" or "ProxyTerminatingEndpoints is purely opt-in"?

It seems to me that making PTE apply to (existing) Cluster services would be an API break, because it would retroactively impose stricter semantics around pod shutdown. Eg, if a pod has no `readinessProbe` then it will never becoming `serving: false`, so if we apply PTE semantics to it, then suddenly we're requiring that it continues to serve traffic correctly until it exits, whereas before it was allowed to stop serving once it got the initial SIGTERM.

> Maybe I am missing something. if we have a REJECT rule we should not be creating conntrack records anyway?

It seems that the kernel creates conntrack records saying "packets like this should be rejected" so that it can reject them immediately in the future rather than having to send them through iptables. (And we have code to delete those records, but it relies on the assumption that we only have a single set of active endpoints, not separate internal and external endpoints.)

> > 4. We could make the proxy set conntrack labels on UDP traffic going through it...
> 
> This sounds good but how? I don't see an iptables option to set that

`-m connlabel --label X --set` (it's in the `iptables-extensions` man page)

This requires reserving bits for our use like we do with mark-masq and mark-drop.

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -521,194 +521,23 @@ func NewFakeProxier(ipt utiliptables.Interface) *Proxier {
-		t.Logf("%q", r)
-	}
-	t.Errorf("%v", msg)
+	// For some reason the metric is off by one
```
`countRules` is correct. It's the metric itself that has the wrong value. If you have 10 rules, the value of the metric will be `11.0`.

It looks like this is because we're counting the `COMMIT` line as a rule. We could fix that, but maybe some people are already correcting for this and we'd break them if we fixed it? Or maybe that doesn't matter becuase metrics aren't APIs?

thockin replied:
I think the metric is buggy.  We should fix the bug

aojea replied:
it is buggy, I didn't think at that time the precision was important , sorry
https://github.com/kubernetes/kubernetes/pull/99653#discussion_r588448513

danwinship replied:
no, it is that metric

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4391,6 +4348,135 @@ COMMIT
+			terminatingFeatureGate: true,
+			endpointslice: &discovery.EndpointSlice{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("%s-1", "svc1"),
```
I think that's just copied from somewhere else...

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4391,6 +4348,135 @@ COMMIT
+				AddressType: discovery.AddressTypeIPv4,
+				Endpoints: []discovery.Endpoint{
+					{
+						// this endpoint won't be used because it's not local,
```
no, we're intentionally including an endpoint that we want it to ignore, to make sure that it ignores it

thockin replied:
Shouldn't this get a REJECT?

danwinship replied:
No, we only generate a REJECT rule if there are no endpoints at all. For a Local traffic policy service with remote endpoints but no local endpoints, we just DROP traffic, because if the LB is lagging slightly behind kube-proxy and forwards a connection to this node after we stop listening, we don't want to cause a "connection refused" to get sent back to the client. So we just drop the packet, and the client will eventually retry its connection to the load balancer, and the load balancer will (hopefully) send to another node.

I can add more comments to the proxier code explaining this if you want?

danwinship replied:
(That was always the expected behavior, and is how the code worked pre-terminating-endpoints, but we didn't have test cases for all parts of it.)

thockin replied:
If we don't REJECT it, the packet falls off the bottom and can actually hit the network, right?  ISTR this was a problem in the past but I don't recall.  Either way, having a timeout seems bad?

aojea replied:
is not dropped by?
> -A KUBE-FW-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 loadbalancer IP" -j KUBE-MARK-DROP

thockin replied:
I missed that.

thockin replied:
It still feels weird to me that we have inconsistent behavior in these cases.  DROP feels rude.  Is there a reason not to reject?

thockin commented:
> _Some_ of the ugliness is because of things that were done to reduce the number of memory allocations per loop...

Yeah.  I am going.to.play with some ideas but it may be problematic to make this significantly different.  I think there's room for improvement, but maybe not as much as I want.

>> I want this ANYWAY. I think it's more correct. Andrew convinced me we could leave it out at first, but maybe not? It's not clear to me if this would be a complete solution, but I don't think it would be an API break?
>
> You want which part? "ProxyTerminatingEndpoints applies to Cluster services too" or "ProxyTerminatingEndpoints is purely opt-in"?

The first.  Terminating endpoints when there are no non-terminating ones.  Being different here has a cost.

> It seems to me that making PTE apply to (existing) Cluster services would be an API break, because it would retroactively impose stricter semantics around pod shutdown. Eg, if a pod has no `readinessProbe` then it will never becoming `serving: false`, so if we apply PTE semantics to it, then suddenly we're requiring that it continues to serve traffic correctly until it exits, whereas before it was allowed to stop serving once it got the initial SIGTERM.

In such a case today we stop accepting connections, but existing ones would survive until the pod stops.  If we made this change, it would allow new connections.  That's the only difference.  I ack it is a difference but I am not sure it is meaningful.  Concrete examples would help

>> Maybe I am missing something. if we have a REJECT rule we should not be creating conntrack records anyway?
>
> It seems that the kernel creates conntrack records saying "packets like this should be rejected" so that it can reject them immediately in the future rather than having to send them through iptables. (And we have code to delete those records, but it relies on the assumption that we only have a single set of active endpoints, not separate internal and external endpoints.)

I just ran a quick test and did not see any contract records for the rejected host?  Can you show a simple repro?

>>> 4. We could make the proxy set conntrack labels on UDP traffic going through it...
>> 
>> This sounds good but how? I don't see an iptables option to set that
>
> `-m connlabel --label X --set` (it's in the `iptables-extensions` man page)

Huh.  I couldn't find it.  In fact, my iptables doesn't have it, so maybe problematic...

> This requires reserving bits for our use like we do with mark-masq and mark-drop.

We have historically not reserved enough for ourselves.  If we go this route, we should claim a few, just in case.

But we need to assess IPVS, too, Right?

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -709,6 +711,87 @@ func errorf(msg string, rules []iptablestest.Rule, t *testing.T) {
 }
 
+// findAllMatches takes an array of lines and a pattern with one parenthesized group, and
+// returns a sorted array of all of the unique matches of the parenthesized group.
```
this is sorted by string size, is this the expectation?

aojea replied:
ic, so you aggregate the lines by the regex, if you match the service name it aggregates by service
you can close this comment

danwinship replied:
no it's not? We insert all the matches into a string set, and then `allMatches.List()` returns the elements sorted as strings...

aojea replied:
yeah, I wrote a nonsense, but I got it after checking how the test uses this function, :sweat_smile:

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1002,7 +998,82 @@ func (proxier *Proxier) syncProxyRules() {
+		readyEndpointChains = readyEndpointChains[:0]
+		localEndpointChains = localEndpointChains[:0]
+		for _, ep := range allEndpoints {
+			epInfo, ok := ep.(*endpointsInfo)
+			if !ok {
+				klog.ErrorS(err, "Failed to cast endpointsInfo", "endpointsInfo", ep)
+				continue
+			}
```
is this needed here? you iterate the same list in line 1005 without casting the endpoint

danwinship replied:
Yes, we need to cast from `proxy.Endpoint` to `*endpointsInfo` in this loop so we can call `epInfo.endpointChain()`.

This is totally ugly and I'd like to get rid of it, but it's hard to do without either (a) moving `endpointChain()` into `proxy.Endpoint` even though it's iptables-specific, or (b) doing more memory allocations.

The `!ok` here is really a "can't happen" though, so in theory we could just make the cast blindly. But I didn't like doing that without adding checks somewhere to make sure that it _really_ can't happen. (It's not possible for a non-`*endpointsInfo` object to end up in that list, but it's not _obvious_ that it's not possible.)

thockin replied:
I think it's because we use the interface types acrss iptables and IPVS, and they have different impl types.  I don't like it either.

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4391,6 +4348,135 @@ COMMIT
+-A KUBE-SERVICES -m comment --comment "ns1/svc1 loadbalancer IP" -m tcp -p tcp -d 10.1.2.3/32 --dport 80 -j KUBE-FW-AQI2S6QIMU7PVVRP
+-A KUBE-FW-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 loadbalancer IP" -j KUBE-XLB-AQI2S6QIMU7PVVRP
+-A KUBE-FW-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 loadbalancer IP" -j KUBE-MARK-DROP
+-A KUBE-XLB-AQI2S6QIMU7PVVRP -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -s 10.0.0.0/24 -j KUBE-SVC-AQI2S6QIMU7PVVRP
```
as Tim says https://github.com/kubernetes/kubernetes/pull/106030/commits/97a93282f7bd4d87b10d6de95ec7db5cc5b24cd7#r742516790 , 
 this will jump to an empty rule and internal traffic will silently dropped by `KUBE-MARK-DROP` ?

aojea replied:
answered here https://github.com/kubernetes/kubernetes/pull/106030#discussion_r742902951

danwinship replied:
I'm not sure exactly what version of the PR you reviewed, but in the current version, there's a commit "`proxy/iptables: test that we create a consistent set of iptables rules`" that ensures that we never jump to a chain that doesn't exist or contains no rules (and the `KUBE-XLB-AQI2S6QIMU7PVVRP` rule here is jumping to `KUBE-SVC-AQI2S6QIMU7PVVRP`, which does exist and does contain rules).

aojea replied:
my fault, I need more coffee
> -A KUBE-SVC-AQI2S6QIMU7PVVRP -m comment --comment "ns1/svc1 cluster IP" -m tcp -p tcp -d 172.20.1.1/32 --dport 80 ! -s 10.0.0.0/24 -j KUBE-MARK-MASQ

s/empty/a noop rule for pod traffic/

danwinship commented:
> In such a case today we stop accepting connections, but existing ones would survive until the pod stops. If we made this change, it would allow new connections. That's the only difference. I ack it is a difference but I am not sure it is meaningful. Concrete examples would help

Currently, it is legal for a pod to close its listening socket when it gets SIGTERM, because it knows kube-proxy won't send it any more connections after that. If we impose terminating endpoints semantics on everyone, then pods are no longer allowed to do that, and any pod that was doing that will cause some client connections to be spuriously rejected.

(Maybe we could say "only pods with `readinessProbe`s get terminating endpoints semantics" and then re-check readiness after sending the SIGTERM, and that would fix all of the problems? Although really, it's currently more-or-less valid for a service to have a `readinessProbe` that returns incorrect answers during termination... We really didn't impose many rules on terminating pods in the past, so who knows what they might be doing?)

But actually, take a step back; why would we want terminating endpoints semantics for Cluster services anyway? The reason it's useful in the Local traffic policy case is because it helps to avoid race conditions. That argument doesn't apply to Cluster traffic policy services; if all of the local endpoints are Terminating but some remote endpoints are Ready, then we can just use the remote endpoints. And if all local _and_ remote endpoints are Terminating, then there is no good reason to accept new connections at all; the Service is supposed to be dead at that point.

danwinship commented:
> But we need to assess IPVS, too, Right?

Yes. (But that applies only to the parts of the discussion that don't get fixed by this PR anyway.)

aojea commented:
> (Maybe we could say "only pods with `readinessProbe`s get terminating endpoints semantics" and then re-check readiness after sending the SIGTERM, and that would fix all of the problems? Although really, it's currently more-or-less valid for a service to have a `readinessProbe` that returns incorrect answers during termination... We really didn't impose many rules on terminating pods in the past, so who knows what they might be doing?)

-1 to complicate this logic even more

> But actually, take a step back; why would we want terminating endpoints semantics for Cluster services anyway? The reason it's useful in the Local traffic policy case is because it helps to avoid race conditions. That argument doesn't apply to Cluster traffic policy services; if all of the local endpoints are Terminating but some remote endpoints are Ready, then we can just use the remote endpoints. And if all local _and_ remote endpoints are Terminating, then there is no good reason to accept new connections at all; the Service is supposed to be dead at that point.

+1 to reduce the scope of terminating to the reported problem of Local and External

thockin commented:
>> In such a case today we stop accepting connections, but existing ones would survive until the pod stops. If we made this change, it would allow new connections. That's the only difference. I ack it is a difference but I am not sure it is meaningful. Concrete examples would help
>
> Currently, it is legal for a pod to close its listening socket when it gets SIGTERM, because it knows kube-proxy won't send it any more connections after that.

That's not actually true.  Kubelet's delivery of SIGTERM is async to kube-proxy's updates, especially in a multi-node cluster.  All it takes is one kube-proxy to be laggy and traffic can keep flowing to that pod.

> If we impose terminating endpoints semantics on everyone, then pods are no longer allowed to do that, and any pod that was doing that will cause some client connections to be spuriously rejected.

I think we already have that.  But remember, we should ONLY use terminating endpoints when there are no non-terminating ones.  As such, if you close your listening socket, it's going to fail either way.  If you don't close your socket, it has a chance to keep working.

> But actually, take a step back; why would we want terminating endpoints semantics for Cluster services anyway?

The biggest argument in my mind is to reduce the behavioral differences between internal and external.  This conntrack issue is a great example.

>> But we need to assess IPVS, too, Right?
>
> Yes. (But that applies only to the parts of the discussion that don't get fixed by this PR anyway.)

My point was that we should consider implementability.  If we can't manage to implement it (e.g. conntrack) with a particular semantic (e.g. terminating endpoints for external but not internal), then that semantic is suspect.

danwinship committed:
proxy/iptables: Fix sync_proxy_rules_iptables_total metric

It was counting the number of lines including the "COMMIT" line at the
end, so it was off by one.

danwinship commented:
repushed with various fixes based on review, plus a fix to the linecount metric (although I wouldn't mind splitting that out into a separate PR because it really doesn't go with the other stuff).

aojea commented:
lgtm

jayunit100 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1002,7 +998,82 @@ func (proxier *Proxier) syncProxyRules() {
+
+		// Scan the endpoints list to see what we have. "hasEndpoints" will be true
+		// if there are any usable endpoints for this service anywhere in the cluster.
+		var hasEndpoints, hasLocalReadyEndpoints, hasLocalServingTerminatingEndpoints bool
```
since these variables are just used to calculate `useTerminatingEndpoints` can we move this into another function? just for understandability

thockin replied:
I think we should do some careful refactoring of all this code (I started some in https://github.com/kubernetes/kubernetes/pull/106158) but we can do those as small followups.  Balancing readability of this code with the desire to DRY is hard.

jayunit100 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1002,7 +998,82 @@ func (proxier *Proxier) syncProxyRules() {
+				continue
+			}
+
+			endpointChain := epInfo.endpointChain(svcNameString, protocol)
```
again, feel free to ignore this but, i think everything below here has side effects... so , it could be separated but ... this is more me just thinking out loud as i try to understand the code

jayunit100 commented:
fixes #106182

thockin reviewed (commented):
Does IPVS get ready/terminating/localterminating right?  Seems like an opportunity to write this once.

thockin commented:
I'm going to LGTM - we can do followups

thockin commented:
/lgtm

aojea commented:
> I'm going to LGTM - we can do followups

we also have to review the in-flight keps, I don't know if it was Dan or you or both who commented it already, but reviewing them I'm really concerned about the future conflicts and interactions and the lots of KEPs that touch kube-proxy https://github.com/orgs/kubernetes/projects/10

alpha Graceful Termination for Local External Traffic Policy https://github.com/kubernetes/enhancements/issues/1669
alpha Support of mixed protocols in Services with type=LoadBalancer https://github.com/kubernetes/enhancements/issues/1435
alpha Topology Aware Hints https://github.com/kubernetes/enhancements/issues/2433
beta Tracking Terminating Endpoints https://github.com/kubernetes/enhancements/issues/1672
beta Service Internal Traffic Policy https://github.com/kubernetes/enhancements/issues/2086
beta Optionally Disable NodePorts for Service Type=LoadBalancer https://github.com/kubernetes/enhancements/issues/1864
beta Service Type=LoadBalancer Class https://github.com/kubernetes/enhancements/issues/1959

robscott commented:
@aojea agree that we need to be careful here, there are a lot of changes coming to kube-proxy and the result could be inconsistencies like this fixes + non-iptables proxiers falling behind. https://github.com/kubernetes/kubernetes/issues/100313 covers the need for hints and traffic policy overlap, and @danwinship did a great job at describing the changes we'd need in that issue. With that said, this is all getting sufficiently complex that it's going to be difficult to keep all proxiers in sync.

aojea commented:
how hard can be merge all the keps and break it down in different subtasks so we can work in different streams without breaking each others?

I can see clearly 2 themes:
- loadbalancer service type changes
- cluster traffic behavior (internal, external, ...)

robscott commented:
At this point @danwinship has volunteered to take on #100313 which I think is really your second bullet point. I'm not sure who's working on the LoadBalancer/NodePort type changes.

aojea commented:
_if Dan is on it we are in good hands_ ;)

jayunit100 commented:
WRT other proxiers.. the userspace and windows proxiers will be alot better off in KPNG long term bc of what rob mentioned. We are always playing catch-up with those .

We're deprecating userspace now and we have MRs in flight for both of these to run in kpng:

I.e. here's our first windows one, https://github.com/kubernetes-sigs/kpng/pull/83

And the userspace https://github.com/kubernetes-sigs/kpng/pull/55

In that case the in tree proxiers could be whittled down to just iptables and IPVS which work quite well and aren't too hard to keep in sync.

aojea commented:
@danwinship until which version do we have to backport this?

danwinship commented:
just 1.22; that's when terminating endpoint support was first added

danwinship commented:
(and only the last commit actually has bugfixes; the rest is all unit test changes, and I'm not sure if they'll backport easily, but it's not really _necessary_ to backport them if they don't)

(well, ok, the "fix the metric" commit is a bugfix too, but it's not worth backporting)

aojea commented:
let me know if you can do it, next friday is the last day for the cherry-pick

> The cherry-pick deadline for the 1.22, 1.21 and 1.20 branches is next Friday, November 12, 2021, EOD PT.

aojea commented:
> let me know if you can do it, next friday is the last day for the cherry-pick
> 
> > The cherry-pick deadline for the 1.22, 1.21 and 1.20 branches is next Friday, November 12, 2021, EOD PT.

cherry-pick to 1.22 https://github.com/kubernetes/kubernetes/pull/106373
